## **PROJECT A**

##A.1



Install cryptography library & necessary classes


In [ ]:
pip install cryptography

In [ ]:
import os
from cryptography.fernet import Fernet
from cryptography.hazmat.primitives.ciphers.aead import ChaCha20Poly1305, AESGCM


Specify the File_path & read the contents of CryptoFile

from Google Drive

In [ ]:
from google.colab import drive
drive.mount('/content/drive')


#FILE ID ON MY OWN DRIVE
file_path='/content/drive/MyDrive/cryptography_project/CryptoFile.pdf'

def read_file(file_path):
  with open(file_path, 'rb') as file:
      content = file.read()
  return content

# Read the content of the CryptoFile.pdf file
CryptoFile_content = read_file(file_path)
print(f"CryptoFile_content size: {len(CryptoFile_content)} bytes.")

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
CryptoFile_content size: 201675 bytes.


Define Fernet encryption Function

Encrypt the PDF content and Print the encrypted_txt Size to compare

In [ ]:
#Define encryption Function (Fernet Encryption)
def encrypt_with_fernet(data):
    key = Fernet.generate_key()
    f = Fernet(key)
    encrypted_txt = f.encrypt(data)
    return key, encrypted_txt

def main(file_path):
    data = read_file(file_path)

    # Encrypt with Fernet
    fernet_key, fernet_encrypted_txt = encrypt_with_fernet(data)
    print(f"Fernet encrypted_txt size: {len(fernet_encrypted_txt)} bytes")

if __name__ == "__main__":
    main(file_path)

Fernet encrypted_txt size: 268984 bytes


Define ChaCha20Poly1305 (RFC 7539) encryption Function

Encrypt the PDF content and Print the encrypted_txt Size to compare

In [ ]:
def encrypt_with_chacha20poly1305(data):

    key = ChaCha20Poly1305.generate_key()
    aad = b"Additional_Encrypt_From_GeorgeK"
    nonce = os.urandom(12)
    chacha = ChaCha20Poly1305(key)
    encrypted_txt = chacha.encrypt(nonce, data, aad)
    return key, nonce, encrypted_txt


def main(file_path):
    data = read_file(file_path)

    # Encrypt with ChaCha20Poly1305
    chacha20_key, chacha20_nonce, chacha20_encrypted_txt = encrypt_with_chacha20poly1305(data)
    print(f"ChaCha20Poly1305 encrypted_txt size: {len(chacha20_encrypted_txt)} bytes")

if __name__ == "__main__":
    main(file_path)

ChaCha20Poly1305 encrypted_txt size: 201691 bytes


Define AES-GCM (RFC 7539) encryption Function

Encrypt the PDF content and Print the encrypted_txt Size to compare

In [ ]:
def encrypt_with_aesgcm(data):
    key = os.urandom(32)
    aesgcm = AESGCM(key)
    aad = b"Additional_Encrypt_From_GeorgeK_for_AES"
    nonce = os.urandom(12)
    encrypted_txt = aesgcm.encrypt(nonce, data, aad)
    return key, nonce, encrypted_txt

def main(file_path):
    data = read_file(file_path)

    # Encrypt with AES-GCM
    aesgcm_key, aesgcm_nonce, aesgcm_encrypted_txt = encrypt_with_aesgcm(data)
    print(f"AES-GCM encrypted_txt size: {len(aesgcm_encrypted_txt)} bytes")

if __name__ == "__main__":
    main(file_path)

AES-GCM encrypted_txt size: 201691 bytes


Conclusions:



*  CryptoFile_content size: 201675 bytes.

*  Fernet encrypted_txt size: 268984 bytes
*  ChaCha20Poly1305 encrypted_txt size: 201691 bytes


*  AES-GCM encrypted_txt size: 201691 bytes





##A.2

Install necessary modules

In [ ]:
from cryptography.hazmat.primitives import cmac, hmac, hashes
from cryptography.hazmat.primitives.ciphers import algorithms
from cryptography.hazmat.primitives.poly1305 import Poly1305

CMAC (Cipher-based Message Authentication Code)

In [ ]:
def calculate_cmac(data, key):
    c = cmac.CMAC(algorithms.AES(key))
    c.update(data)
    cmac_signature = c.finalize()
    return cmac_signature


def main(file_path):
    data = read_file(file_path)

    # Calculate CMAC value
    cmac_signature = calculate_cmac(data, os.urandom(32))
    print(f"CMAC value size: {len(cmac_signature)} bytes")
    print(cmac_signature)

if __name__ == "__main__":
    main(file_path)


CMAC value size: 16 bytes
b'\xb5\xecF\x9e\x8f\xe1\xe6Q@K#\xd3\xd1\x94>j'


HMAC(Hash-based message authentication codes)

In [ ]:
def calculate_hmac(data, key):
    h = hmac.HMAC(key, hashes.SHA256())
    h.update(data)
    signature = h.finalize()
    signature
    return signature


def main(file_path):
    data = read_file(file_path)

    # Calculate HMAC
    hmac_value = calculate_hmac(data, os.urandom(32))
    print(f"HMAC value size: {len(hmac_value)} bytes")
    print(hmac_value)

if __name__ == "__main__":
    main(file_path)

HMAC value size: 32 bytes
b'4E\x99\x82S\r@#\xbf\xa5E\xde^f\x1a\xd2\xb8\xfc\x86S\x86\xbf.\xc0V\xd6\xad\xa8\xfb\xc2Q\xfe'


Poly1305 authenticator using a 32-byte key to produce 16-byte tag

In [ ]:
def calculate_poly1305(data, key):
    p = Poly1305(key)
    p.update(data)
    poly1305_auth = p.finalize()
    return poly1305_auth


def main(file_path):
    data = read_file(file_path)


    # Calculate Poly1305
    poly1305_value = calculate_poly1305(data, os.urandom(32))
    print(f"Poly1305 tag size: {len(poly1305_value)} bytes")
    print(poly1305_value)

if __name__ == "__main__":
    main(file_path)


Poly1305 tag size: 16 bytes
b'\xbc9\x9d2\x1eN,$\x86\x9dz\x12\x13\xc0\xde{'


Conclusions:



*  CMAC value size: 16 bytes

*  HMAC value size: 32 bytes
*  Poly1305 tag size: 16 bytes





##A.3

extra Install necessary modules

In [ ]:
from cryptography.hazmat.primitives.asymmetric import dsa, rsa, padding
from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey

DSA public-key algorithm for signing messages

In [ ]:
def sign_dsa(data, private_key):
  signature = private_key.sign(data, hashes.SHA256())
  return signature


def main(file_path):
    data = read_file(file_path)

    # Sign with DSA key generated
    private_key = dsa.generate_private_key(key_size=2048)
    dsa_signature = sign_dsa(data, private_key)
    print(f"DSA signature size: {len(dsa_signature)} bytes")
    print(dsa_signature)

if __name__ == "__main__":
    main(file_path)


DSA signature size: 70 bytes
b'0D\x02 \x10\xa9\xfa\x12p\xcb\xac\xb7\x90T\x1d.P{S\xb9\xe9\xb7\xa2J\x940\xb9!D\xeb\xf9_\xd8\x97\xaaH\x02 &\xa4\x87\\^>\xfa\xf7\x07k\x85\x1e@\xdd\x06\xe5t\xb1Z0\xd6\x9a\xd5JU[?\x9f\xfe\xf0\xa9\xe2'


RSA

The RSA algorithm involves four steps: key generation, key distribution, encryption, and decryption.

In [ ]:
def sign_rsa(data, private_key):
    signature = private_key.sign(data,padding.PSS(mgf=padding.MGF1(hashes.SHA256()), salt_length=padding.PSS.MAX_LENGTH), hashes.SHA256())
    return signature


def main(file_path):
    data = read_file(file_path)

    # Sign with RSA
    rsa_private_key = rsa.generate_private_key(public_exponent=65537,key_size=2048)
    rsa_signature = sign_rsa(data, rsa_private_key)
    print(f"RSA signature size: {len(rsa_signature)} bytes")
    print(rsa_signature)

if __name__ == "__main__":
    main(file_path)

RSA signature size: 256 bytes
b'\xd6\xf8\x8ca\x0e\xc3\xbc\xd8z\x93\xa4\x87^\xac\xdb\xbbsGR"{\x16\xfca:\xbf!\xd3g\xc2\xc0\xc1\x18\x7f\x81\xdb!C1\x1c`\xf9\t\x13\x16@z\xd6\xbc\xbb\xf3\xdf\xdbUd\xe63F6z\xe9\xef:\xe7\x18\xda\xe2\xbfQ\xa5\x9b\x9d\xd4e\xe2`\x85\xfe\xfd\x1b\xf2\xcd\xe8\xb3\xeb\xc2\x89\xf8\xb1P\x02\xe0>\xab(\\\x0eI0l\x88]l\xc2\xd8\xf229\rb(\xae\x99u\x9e\xac&/c\t\x0b\x1c\x05#|\xfbI>\x9e`\xac\xb9m\xb0\xef\x02r\xdf\xa28\xe8\x1dn<\xdc>\xdf\xd1\x0cC\x80Wm\xe0\xbb\x0c\xb8\x9bT\x13\xc4\xe7\xa8\x96F!\xc2\x80\x1a\xeb&\xc2\xd4n\x03s\xe90\xde<\xae\xd7K\xc4\x8a\xbc1A\xcc\xa4o\x0cJ\x15\xf5\x0cP\x11\x9f-?\xbb\x1a`\x06t\xd0\x1b\xffo-d1l\x1dj\x0e\xcac\xc1A@\x03\xb8\x89\xdf\xca\x951}\x99\x84Ju\xa2&\xe49\xfa\x15bl\x83x\x96\xd3q#C\xc5\x07W\xa2\x0c\xdb\xae'


Ed25519 signing

In [ ]:
def sign_ed25519(data, private_key):
    signature = private_key.sign(data)
    return signature

def main(file_path):
    data = read_file(file_path)


 # Sign with Ed25519
    private_key = Ed25519PrivateKey.generate()
    ed25519_signature = sign_ed25519(data, private_key)
    public_key = private_key.public_key()

    # Raises InvalidSignature if verification fails
    public_key.verify(ed25519_signature, data)
    print(f"Ed25519 signature size: {len(ed25519_signature)} bytes")
    print(ed25519_signature)

if __name__ == "__main__":
    main(file_path)

Ed25519 signature size: 64 bytes
b'\xb9\xc4ND\xb8\xe2\x80\xd2\xfd\x0e\xb5\xc9\xfe\xcf\xe4\x03\xc3fO\x0f\x13D\xd3\xb2\x1cI\xf5\x19\xa4t\xaa\xf6\xd6\xd9\xc7\x15jt1\x04\xb5\xcc?\xed\xe2\x95\xa2\r\xc8I\xc3j\xcc\xeb!\xcd\x92\xe7\x05u\xc3\xc3!\x08'
